daemon/cmd/kube_proxy_replacement.go

```diff
@@ -233,28 +235,36 @@ func initKubeProxyReplacementOptions() (strict bool) {
 			option.Config.EnableHostServicesPeer = false
 		}
 
-		if option.Config.EnableHostServicesTCP &&
-			(option.Config.EnableIPv4 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET4_CONNECT) != nil ||
-				option.Config.EnableIPv6 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET6_CONNECT) != nil) {
-			msg := "BPF host reachable services for TCP needs kernel 4.17.0 or newer."
-			if strict {
-				log.Fatal(msg)
-			} else {
-				option.Config.EnableHostServicesTCP = false
-				log.Warn(msg + " Disabling the feature.")
-			}
-		}
-		if option.Config.EnableHostServicesUDP &&
-			(option.Config.EnableIPv4 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_UDP4_RECVMSG) != nil ||
-				option.Config.EnableIPv6 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_UDP6_RECVMSG) != nil) {
-			msg := fmt.Sprintf("BPF host reachable services for UDP needs kernel 4.19.57, 5.1.16, 5.2.0 or newer. If you run an older kernel and only need TCP, then specify: --%s=tcp and --%s=%s", option.HostReachableServicesProtos, option.KubeProxyReplacement, option.KubeProxyReplacementPartial)
-			if strict {
-				log.Fatal(msg)
-			} else {
-				option.Config.EnableHostServicesUDP = false
-				log.Warn(msg + " Disabling the feature.")
-			}
+		if option.Config.EnableHostServicesTCP {
+			bpf.TestIPDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET4_CONNECT, bpf.BPF_CGROUP_INET6_CONNECT, func(ipv4Result error, ipv6Result error) {
+				msg := fmt.Sprintf("Test Dummy Prog Faild: [ IPv4: %v, IPv6: %v ], BPF host reachable services for TCP needs kernel 4.17.0 or newer.", ipv4Result, ipv6Result)
```

Comment:
Getting there. I think we could probably get away with something a bit simpler: If the first prog attempt fails with `unix.EPERM` then it probably means that all subsequent loads are also going to fail with the same error, regardless of IPv4/IPv6 or which particular hook?

Additionally, in Cilium's logging we aim to properly use [Structured logging](https://stackify.com/what-is-structured-logging-and-why-developers-need-it/) so that fields are properly propagated in logging systems that are intelligent enough to use them (particularly when formatting messages to JSON logging platforms). What does this mean for the PR here? Rather than combining static content (the message) with dynamic logging content (The `%v` bits), the dynamic content is separately specified into the logging system. In practice, the syntax for this looks like:

```
log.WithError(err).Warn("Cilium cannot load bpf programs. Security profiles like SELinux may be restricting permissions.")
```

If there are additional details you want to share in the logs, you can also add `WithFields(...)`, like `log.WithError(err).WithFields(logfields.IPv4, ipAddr).Warn(...)`.

Reply:
I use `%v` beacause if use `%s` when error is `nil` the message will display like `%!s(<nil>)`, use `%v` don't have this problem.
I was going to use `log.WithError` first time. But it have a problem, when prog check faild. We should display the error detail on the log,  `log.WithError` only accept one args. But I don't known witch check is faild, `ipv6` or `ipv4`? If I add the code to decide whitch one is faild. It will be too complex. So I add the error detail in the log content.

Reply:
done